### ISSUE TITLE
Dependency based redecompilation

### ISSUE THREAD

**Issue by sad-dev** [NationalSecurityAgency/ghidra#2424]
**Is your feature request related to a problem? Please describe.**

Any change to the state of the database causes a flushing of the decompilation cache. This is inefficient and a major pain point on large binaries and/or large functions. 

https://github.com/NationalSecurityAgency/ghidra/issues/1883
https://github.com/NationalSecurityAgency/ghidra/issues/1871

**Describe the solution you'd like**
Certain changes should instead lead to a selective recompilation (upon visit) of functions. The obvious candidates are:

a.) Local variables -> should only affect the current function. As mentioned in https://github.com/NationalSecurityAgency/ghidra/issues/1871, directly modifying the decompile tree is theoretically possible, although as an initial version I would not mind just re-decompiling the current function

b.) Changes in the function prototype or parameter renames should change the current function and all decompiled callers of the function

c.) Mapping which functions have decompiled output dependent on a type/structure definition is probably a stretch goal, but could be possible (similar to the find use of type/field feature already in ghidra).

**Describe alternatives you've considered**
I have resorted to using reflection from scripts to directly interact with the decompiler cache while performing the changes outlined in (a) and (b).

**Comment by astrelsky** [comment]
I would like to add that any change to the state of one program's database causes the decompiler cache to be flushed for ALL open programs. This a royal pain when manually comparing programs.

**Comment by gtackett** [comment]
I'm seriously considering 😄 creating a dozen or so extra github logins, just so that I could give this a few more thumbs up.

**Comment by pj58943** [comment]
This is a big issue for large functions -- I have some pathological cases that take several minutes to decompile, and having to wait every time I add a comment or rename a variable can get frustrating.

I'm experimenting with ways to fix this, at least for my own use (as a starting point, I've modified `DecompileProvider.domainObjectChanged` to ignore events that surely will never affect decompilation, such as saving!) -- but it would be good to know if the Ghidra team have any work ongoing in this area, or opinions on implementation strategies.

Ultimately I don't think a perfect solution is possible with the current architecture: undo/redo actions only send DO_OBJECT_RESTORED which inherently requires a full refresh.  But the situation can certainly be improved!

**Comment by dragonmacher** [comment]
A simple approach we have taken with other widgets is to mark them as potential being out-of-date.   This, along with a refresh button would prevent excessive work while still providing some signal that the data may need to be updated.

**Comment by pj58943** [comment]
Good point, that would be a quick and cheap fix for the main cause of frustration!  Probably with a configurable function-complexity threshold, so we don't lose the nice instant feedback in the common case where the decompilation time is not noticeable.  (Or keep a record of the time taken to decompile the function the first time, and use that to decide?)

In my use case I would still like to have at least comments and renames show up instantly in large functions.  I have been experimenting with scripts that use a lot of reflection to modify the CCodeMarkup tree, which seems to be viable in principle.  I don't think it's possible to implement cleanly in a script or extension, though, so doing this properly would involve adding some special-case handling to the core decompiler plugin.

**Comment by dragonmacher** [comment]
>  I don't think it's possible to implement cleanly in a script or extension, though, so doing this properly would involve adding some special-case handling to the core decompiler plugin.

The devil is in the details.  It is certainly possible to update the existing fields, but that would require extra logic to listen to and process events to detect those changes and update accordingly.   This is an interesting suggestion that goes with the 'stale' concept mentioned above.
